File: project/.sdk/tm/lua/utility/prepare_auth.lua
-- ProjectName SDK utility: prepare_auth

local vs = require("utility.struct.struct")

local HEADER_AUTH = "authorization"
local OPTION_APIKEY = "apikey"
local NOT_FOUND = "__NOTFOUND__"

local function prepare_auth_util(ctx)
  local spec = ctx.spec
  if spec == nil then
    return nil, ctx:make_error("auth_no_spec",
      "Expected context spec property to be defined.")
  end

  local headers = spec.headers
  local options = ctx.client:options_map()

  -- Public APIs that need no auth omit the options.auth block entirely.
  if options.auth == nil then
    headers[HEADER_AUTH] = nil
    return spec, nil
  end

  local apikey = vs.getprop(options, OPTION_APIKEY, NOT_FOUND)

  if apikey == nil
    or (type(apikey) == "string" and (apikey == NOT_FOUND or apikey == ""))
  then
    headers[HEADER_AUTH] = nil
  else
    local auth_prefix = ""
    local ap = vs.getpath(options, "auth.prefix")
    if type(ap) == "string" then
      auth_prefix = ap
    end
    local apikey_val = ""
    if type(apikey) == "string" then
      apikey_val = apikey
    end
    headers[HEADER_AUTH] = auth_prefix .. " " .. apikey_val
  end

  return spec, nil
end

return prepare_auth_util

File: project/.sdk/tm/php/utility/PrepareAuth.php
<?php
declare(strict_types=1);

// ProjectName SDK utility: prepare_auth

class ProjectNamePrepareAuth
{
    private const HEADER_AUTH = 'authorization';
    private const OPTION_APIKEY = 'apikey';
    private const NOT_FOUND = '__NOTFOUND__';

    public static function call(ProjectNameContext $ctx): array
    {
        $spec = $ctx->spec;
        if (!$spec) {
            return [null, $ctx->make_error('auth_no_spec', 'Expected context spec property to be defined.')];
        }

        $headers = &$spec->headers;
        $options = $ctx->client->options_map();
        $apikey = \Voxgig\Struct\Struct::getprop($options, self::OPTION_APIKEY, self::NOT_FOUND);

        if (is_string($apikey) && $apikey === self::NOT_FOUND) {
            unset($headers[self::HEADER_AUTH]);
        } else {
            $auth_prefix = \Voxgig\Struct\Struct::getpath($options, 'auth.prefix') ?? '';
            $apikey_val = is_string($apikey) ? $apikey : '';
            $headers[self::HEADER_AUTH] = "{$auth_prefix} {$apikey_val}";
        }

        return [$spec, null];
    }
}
